src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -1896,29 +1896,43 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
-                    // try to obtain a type of a local when it is used before declaration, we use
-                    // a special error type symbol. However, semantic model should return the same
-                    // type information for usage of a local before and after its declaration.
-                    // We will detect the use before declaration cases and replace the error type
-                    // symbol with the one obtained from the local. It should be safe to get the type
-                    // from the local at this point.
-                    if (type.IsErrorType() && boundExpr.Kind == BoundKind.Local)
+                    switch (boundExpr.Kind)
                     {
-                        var extended = type as ExtendedErrorTypeSymbol;
-                        if ((object)extended != null && extended.VariableUsedBeforeDeclaration)
-                        {
-                            type = ((BoundLocal)boundExpr).LocalSymbol.Type;
-                        }
-                    }
+                        case BoundKind.Local:
+                            {
+                                // Use of local before declaration requires some additional fixup.
+                                // Due to complications around implicit locals and type inference, we do not
+                                // try to obtain a type of a local when it is used before declaration, we use
+                                // a special error type symbol. However, semantic model should return the same
+                                // type information for usage of a local before and after its declaration.
+                                // We will detect the use before declaration cases and replace the error type
+                                // symbol with the one obtained from the local. It should be safe to get the type
+                                // from the local at this point.
+                                if (type is ExtendedErrorTypeSymbol extended && extended.VariableUsedBeforeDeclaration)
+                                {
+                                    type = ((BoundLocal)boundExpr).LocalSymbol.Type;
+                                }
+                                break;
+                            }
+                        case BoundKind.ConvertedTupleLiteral:
+                            {
+                                // The bound tree always fully binds tuple literals. From the language point of
+                                // view, however, converted tuple literals represent tuple conversions
+                                // from tuple literal expressions which may or may not have types
+                                type = ((BoundConvertedTupleLiteral)boundExpr).NaturalTypeOpt;
+                                break;
+                            }
+                        case BoundKind.StackAllocArrayCreation:
+                            {
```

Comment:
This is within the code that tests HasExpressionType. But a stackalloc expression shouldn't have a type. This is confusing.